### ISSUE TITLE
Update fails with two different error codes (HASSIO 0.83.2)

### ISSUE THREAD

**Issue by kegra** [NeoAcheron/midea-ac-py#5]
This error is not that frequent:

Update for climate.midea_10995116286482 fails
Traceback (most recent call last):
File “/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py”, line 221, in async_update_ha_state
await self.async_device_update()
File “/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py”, line 349, in async_device_update
await self.hass.async_add_executor_job(self.update)
File “/usr/local/lib/python3.6/concurrent/futures/thread.py”, line 56, in run
result = self.fn(*self.args, **self.kwargs)
File “/config/custom_components/climate/midea.py”, line 80, in update
self._device.refresh()
File “/config/deps/lib/python3.6/site-packages/midea/device.py”, line 82, in refresh
data = self._cloud_client.appliance_transparent_send(self.id, data)
File “/config/deps/lib/python3.6/site-packages/midea/cloud.py”, line 124, in appliance_transparent_send
‘applianceId’: id
File “/config/deps/lib/python3.6/site-packages/midea/cloud.py”, line 62, in api_request
raise ValueError(response[‘errorCode’], response[‘msg’])
ValueError: (‘3176’, 'The asyn reply does not exist ')

this error is the most frequent one:
Update for climate.midea_10995116286482 fails
Traceback (most recent call last):
File “/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py”, line 221, in async_update_ha_state
await self.async_device_update()
File “/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py”, line 349, in async_device_update
await self.hass.async_add_executor_job(self.update)
File “/usr/local/lib/python3.6/concurrent/futures/thread.py”, line 56, in run
result = self.fn(*self.args, **self.kwargs)
File “/config/custom_components/climate/midea.py”, line 80, in update
self._device.refresh()
File “/config/deps/lib/python3.6/site-packages/midea/device.py”, line 82, in refresh
data = self._cloud_client.appliance_transparent_send(self.id, data)
File “/config/deps/lib/python3.6/site-packages/midea/cloud.py”, line 124, in appliance_transparent_send
‘applianceId’: id
File “/config/deps/lib/python3.6/site-packages/midea/cloud.py”, line 62, in api_request
raise ValueError(response[‘errorCode’], response[‘msg’])
ValueError: (‘3106’, ‘invalidSession’)

**Comment by NeoAcheron** [comment]
I've added some retry logic and better error handling for those two errors, they shouldn't appear that often anymore. 

I'll keep this issue open for a little while longer to make sure there aren't any other errors that need to be handled in the same way.

**Comment by kegra** [comment]
seems fine , will run few hours or more and report back , thank you devoting time and energy for the community.

**Comment by kegra** [comment]
after few hours of running got this:

Update for climate.midea_10995116286482 fails
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py", line 221, in async_update_ha_state
    await self.async_device_update()
  File "/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py", line 349, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/climate/midea.py", line 85, in update
    self._device.refresh()
  File "/config/deps/lib/python3.6/site-packages/midea/device.py", line 140, in refresh
    data = self._cloud_service.appliance_transparent_send(self.id, data)
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 137, in appliance_transparent_send
    'applianceId': id
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 68, in api_request
    return self.api_request(endpoint, args)
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 62, in api_request
    self.handle_api_error(int(response['errorCode']), response['msg'])
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 165, in handle_api_error
    handler()
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 153, in throw
    raise ValueError(error_code, message)
ValueError: (9999, 'system error')

just reporting so please don't hate me :-)

**Comment by NeoAcheron** [comment]
lol thanks, no hate from me, I'm doing this for my own benefit as well, I want to get my AC to be controlled reliably! I might as well make it public and let everyone else in on the fun!

That is an interesting error! 

I'll add more error handling tonight, I'm not even sure if these can be retried. The cloud API isn't very robust, it seems to be extremely finicky. The app isn't better, a lot of people on the Google Play store complaining about issues.

In addition I also got more errors this morning:
```
Update for climate.midea_15393162795035 fails
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py", line 221, in async_update_ha_state
    await self.async_device_update()
  File "/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity.py", line 349, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/climate/midea.py", line 85, in update
    self._device.refresh()
  File "/config/deps/lib/python3.6/site-packages/midea/device.py", line 140, in refresh
    data = self._cloud_service.appliance_transparent_send(self.id, data)
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 137, in appliance_transparent_send
    'applianceId': id
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 62, in api_request
    self.handle_api_error(int(response['errorCode']), response['msg'])
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 165, in handle_api_error
    handler()
  File "/config/deps/lib/python3.6/site-packages/midea/cloud.py", line 153, in throw
    raise ValueError(error_code, message)
ValueError: (3004, 'value is illegal.')```

**Comment by NeoAcheron** [comment]
I've added more codes to handle, but it seemed like there would be constant errors, because HA tried to update both my AC's at the same time. The cloud API doesn't like that at all. I made a change to have the update happen async instead, which allows it to hold back the one device while the other is updated. It seems to have reduced the errors, but there is still a potential for a race condition. 

I'm working on having the cloud client manage a lock to call devices, so that only one at a time can be modified. This will impact performance slightly, but it will make things much more stable.
